write-fonts/src/tables/glyf.rs

```diff
@@ -86,29 +88,291 @@ pub enum BadKurbo {
+            continue;
+        }
+        let prv = if i > 0 { i - 1 } else { points.len() - 1 };
+        let nxt = if i < points.len() - 1 { i + 1 } else { 0 };
+        let p0 = &points[prv];
+        let p2 = &points[nxt];
+        if p0.on_curve || p0.on_curve != p2.on_curve {
+            continue;
+        }
+        // if the distance between p1 and p0 is approximately the same as the distance
+        // between p2 and p1, then we can drop p1
+        let p1p0 = p1.distance(p0);
+        let p2p1 = p2.distance(p1);
+        if (p1p0 - p2p1).abs() < tolerance {
+            result.insert(i);
+        }
+    }
+    result
 }
 
-impl OtPoint for kurbo::Point {
-    fn get(self) -> (i16, i16) {
-        (self.x.ot_round(), self.y.ot_round())
+// The MultiZip is adapted from https://stackoverflow.com/a/55292215
+
+/// Iterator that iterates over a vector of iterators simultaneously
+struct MultiZip<I: Iterator>(Vec<I>);
+
+impl<I: Iterator> Iterator for MultiZip<I> {
+    type Item = Vec<I::Item>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.iter_mut().map(Iterator::next).collect()
     }
 }
 
-impl OtPoint for (i16, i16) {
-    fn get(self) -> (i16, i16) {
-        self
+pub fn simple_glyphs_from_kurbo(paths: &[BezPath]) -> Result<Vec<SimpleGlyph>, BadKurbo> {
+    // check that all paths have the same number of elements so we can zip them together
+    let num_elements: Vec<usize> = paths.iter().map(|path| path.elements().len()).collect();
+    if num_elements.iter().any(|n| *n != num_elements[0]) {
+        return Err(BadKurbo::UnequalNumberOfElements(num_elements));
+    }
+    let path_iters = MultiZip(paths.iter().map(|path| path.iter()).collect());
+    let mut contours: Vec<InterpolatableContourBuilder> = Vec::new();
+    let mut current: Option<InterpolatableContourBuilder> = None;
+    let num_glyphs = paths.len();
+    for (i, elements) in path_iters.enumerate() {
+        // check that all i-th path elements have the same types
+        let mut el_types: Vec<&str> = Vec::with_capacity(num_glyphs);
+        el_types.extend(elements.iter().map(|el| match el {
+            kurbo::PathEl::MoveTo(_) => "M",
+            kurbo::PathEl::LineTo(_) => "L",
+            kurbo::PathEl::QuadTo(_, _) => "Q",
+            kurbo::PathEl::CurveTo(_, _, _) => "C",
+            kurbo::PathEl::ClosePath => "Z",
+        }));
+        if el_types.iter().any(|t| *t != el_types[0]) {
+            return Err(BadKurbo::InconsistentPathElements(i, el_types));
+        }
+
+        // elements is never empty (if it were, MultiZip would have stopped), hence the unwrap
+        let first_el = elements.first().unwrap();
+        match first_el {
+            kurbo::PathEl::MoveTo(_) => {
+                // we have a new contour, flush the current one
+                if let Some(prev) = current.take() {
+                    contours.push(prev);
+                }
+                let mut pts = Vec::with_capacity(num_glyphs);
```

Comment:
If the `new`, `line_to` and `quad_to` methods are changed to take slices, these allocations can be reused. I don’t want to stumble into premature optimization but we did have a recent patch to address malloc in a profile and these per point allocations are likely to be high frequency.

Reply:
No worries! You captured the bulk of it. I’ll do a PR against this branch demonstrating what I mean about reusing the vecs.

Reply:
Hopefully #378 adds some clarity!